samza-core/src/main/java/org/apache/samza/container/LocalityManager.java

```diff
@@ -93,35 +124,30 @@ public LocalityManager(CoordinatorStreamManager coordinatorStreamManager) {
    * @param jmxAddress  the JMX URL address
    * @param jmxTunnelingAddress  the JMX Tunnel URL address
    */
-  public void writeContainerToHostMapping(String containerId, String hostName, String jmxAddress,
-      String jmxTunnelingAddress) {
-    if (coordinatorStreamManager == null) {
-      throw new IllegalStateException("No coordinator stream manager to write locality info to.");
-    }
-
+  public void writeContainerToHostMapping(String containerId, String hostName, String jmxAddress, String jmxTunnelingAddress) {
     Map<String, String> existingMappings = containerToHostMapping.get(containerId);
-    String existingHostMapping =
-        existingMappings != null ? existingMappings.get(SetContainerHostMapping.HOST_KEY) : null;
+    String existingHostMapping = existingMappings != null ? existingMappings.get(SetContainerHostMapping.HOST_KEY) : null;
     if (existingHostMapping != null && !existingHostMapping.equals(hostName)) {
       LOG.info("Container {} moved from {} to {}", new Object[]{containerId, existingHostMapping, hostName});
     } else {
       LOG.info("Container {} started at {}", containerId, hostName);
     }
-    coordinatorStreamManager.send(
-        new SetContainerHostMapping(CONTAINER_PREFIX + containerId, String.valueOf(containerId), hostName, jmxAddress,
-            jmxTunnelingAddress));
+
+    metadataStore.put(keySerde.toBytes(containerId), valueSerde.toBytes(hostName));
+
     Map<String, String> mappings = new HashMap<>();
     mappings.put(SetContainerHostMapping.HOST_KEY, hostName);
     mappings.put(SetContainerHostMapping.JMX_URL_KEY, jmxAddress);
     mappings.put(SetContainerHostMapping.JMX_TUNNELING_URL_KEY, jmxTunnelingAddress);
     containerToHostMapping.put(containerId, mappings);
```

Comment:
I don't see the purpose of this as well. It looks like `writeContainerToHostMapping` seems to be invoked only during the start of the container and this update is not even read subsequently. I am surprised as to how this is working prior to your change since the old implementation sets `containerToHostMapping` to `unmodifiableMap` inside `readContainerLocality`

Reply:
I should have been more explicit. I do understand the need for it to be in coordinator stream. I was referring to the part on why we persist the information in a local map within `LocalityManager`.
`  Map<String, String> mappings = new HashMap<>();
    mappings.put(SetContainerHostMapping.HOST_KEY, hostName);
    mappings.put(SetContainerHostMapping.JMX_URL_KEY, jmxAddress);
    mappings.put(SetContainerHostMapping.JMX_TUNNELING_URL_KEY, jmxTunnelingAddress);
    containerToHostMapping.put(containerId, mappings);
`
After writing the locality information to the coordinator stream as part of the startup, the locality information is not accessed for the rest of the lifecycle of the running container. Subsequent reads happen only for container movements, restarts or deploy where it is guaranteed to see the update from coordinator stream.